# Generated by Django 2.2.14 on 2020-12-04 11:00

from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ('polaaar', '0045_auto_20201204_1051'),
    ]

    operations = [
        migrations.AlterField(
            model_name='samplemetadata',
            name='geo_loc_name',
            field=models.CharField(blank=True, help_text='The geographical origin of the sample as defined by the country or sea name followed by specific region name. Country or sea names should be chosen from the INSDC country list (http://insdc.org/country.html); or the GAZ ontology (v 1.512) (http://purl.bioontology.org/ontology/GAZ)', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='samplemetadata',
            name='investigation_type',
            field=models.CharField(blank=True, help_text='Nucleic Acid Sequence Report is the root element of all MIGS/MIMS compliant reports as standardized by Genomic Standards Consortium. This field is either eukaryote,bacteria,virus,plasmid,organelle, metagenome,mimarks-survey, or mimarks-specimen', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='samplemetadata',
            name='locality',
            field=models.CharField(blank=True, help_text='http://rs.tdwg.org/dwc/terms/locality', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='samplemetadata',
            name='ref_biomaterial',
            field=models.CharField(blank=True, help_text='Primary publication if isolated before genome publication; otherwise, primary genome report', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='samplemetadata',
            name='rel_to_oxygen',
            field=models.CharField(blank=True, help_text='Is this organism an aerobe, anaerobe? Please note that aerobic and anaerobic are valid descriptors for microbial environments', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='samplemetadata',
            name='rightsHolder',
            field=models.CharField(blank=True, help_text='http://purl.org/dc/terms/rightsHolder', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='samplemetadata',
            name='samp_collect_device',
            field=models.CharField(blank=True, help_text='The method or device employed for collecting the sample', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='samplemetadata',
            name='samp_store_loc',
            field=models.CharField(blank=True, help_text='location at which sample was stored; usually name of a specific freezer/room', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='samplemetadata',
            name='source_mat_id',
            field=models.CharField(blank=True, help_text="A unique identifier assigned to a material sample (as defined by http://rs.tdwg.org/dwc/terms/materialSampleID, and as opposed to a particular digital record of a material sample) used for extracting nucleic acids, and subsequent sequencing. The identifier can refer either to the original material collected or to any derived sub-samples. The INSDC qualifiers /specimen_voucher, /bio_material, or /culture_collection may or may not share the same value as the source_mat_id field. For instance, the /specimen_voucher qualifier and source_mat_id may both contain 'UAM:Herps:14' , referring to both the specimen voucher and sampled tissue with the same identifier. However, the /culture_collection qualifier may refer to a value from an initial culture (e.g. ATCC:11775) while source_mat_id would refer to an identifier from some derived culture from which the nucleic acids were extracted (e.g. xatc123 or ark:/2154/R2).", max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='sequences',
            name='primerName_forward',
            field=models.CharField(blank=True, help_text='http://data.ggbn.org/schemas/ggbn/terms/primerNameForward', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='sequences',
            name='primerName_reverse',
            field=models.CharField(blank=True, help_text='http://data.ggbn.org/schemas/ggbn/terms/primerNameReverse', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='sequences',
            name='run_type',
            field=models.CharField(blank=True, help_text='The type of sequencing run performed. E.g. Illumina MiSeq 250bp paired-end', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='sequences',
            name='seqData_accessionNumber',
            field=models.CharField(blank=True, help_text='An assocated INSDC GenBank accession number.', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='sequences',
            name='seqData_projectNumber',
            field=models.CharField(blank=True, help_text='An assocated INSDC BioProject number.', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='sequences',
            name='seqData_runNumber',
            field=models.CharField(blank=True, help_text='An assocated INSDC run accession number. (ERR number)', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='sequences',
            name='seqData_sampleNumber',
            field=models.CharField(blank=True, help_text='An assocated INSDC BioSample number.', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='sequences',
            name='subspecf_gen_lin',
            field=models.CharField(blank=True, help_text='This should provide further information about the genetic distinctness of this lineage by recording additional information i.e biovar, serovar, serotype, biovar, or any relevant genetic typing schemes like Group I plasmid. It can also contain alternative taxonomic information', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='sequences',
            name='target_gene',
            field=models.CharField(blank=True, help_text='Targeted gene or locus name for marker gene studies', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='sequences',
            name='target_subfragment',
            field=models.CharField(blank=True, help_text='Name of subfragment of a gene or locus. Important to e.g. identify special regions on marker genes like V6 on 16S rRNA', max_length=500, null=True),
        ),
        migrations.AlterField(
            model_name='sequences',
            name='type',
            field=models.CharField(blank=True, help_text='The type of study; either genomic; metagenomic; transcriptomic; metatranscriptomic; viral RNA; synthetic or other.', max_length=500, null=True),
        ),
    ]
